sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py

```diff
@@ -755,8 +571,16 @@ def upload_blob(
                 :dedent: 12
                 :caption: Upload a blob to the container.
         """
-        options = self._upload_blob_options(
-            data,
+        if self.require_encryption and not self.key_encryption_key:
+            raise ValueError("Encryption required but no key was provided.")
+        if kwargs.get('cpk') and self.scheme.lower() != 'https':
+            raise ValueError("Customer provided encryption key must be used over HTTPS.")
+        options = _upload_blob_options(
+            encryption_options=self._build_encryption_options(),
+            config=self._config,
+            sdk_moniker=self._sdk_moniker,
+            client=self._client,
+            data=data,
```

Comment:
This is going to seem super nitpicky but, can I suggest a new order to these parameters? Not sure why but the current ordering just feels wrong. Sorry as I know this means changing the helper too :)

```py
data=data,
blob_type=blob_type,
length=length,
metadata=metadata,
encryption_options={
    'required': self.require_encryption,
    'version': self.encryption_version,
    'key': self.key_encryption_key,
    'resolver': self.key_resolver_function
},
config=self._config,
sdk_moniker=self._sdk_moniker,
client=self._client,
**kwargs
```